src/main/java/io/tiledb/java/api/Array.java

```diff
@@ -615,6 +615,179 @@ public Pair getNonEmptyDomainFromName(String name) throws TileDBError {
+   *
+   * @param index index to retrieve metadata from
+   * @return a pair, key and the metadata
+   * @throws TileDBError A TileDB exception
+   */
+  public Pair<String, NativeArray> getMetadataFromIndex(long index) throws TileDBError {
+    return getMetadataFromIndex(BigInteger.valueOf(index));
+  }
+
+  /**
+   * Gets a metadata item from an open array using an index. The array must be opened in READ mode,
+   * otherwise the function will error out.
+   *
+   * @param index index to retrieve metadata from
+   * @return a pair, key and the metadata
+   * @throws TileDBError A TileDB exception
+   */
+  public Pair<String, NativeArray> getMetadataFromIndex(BigInteger index) throws TileDBError {
+    checkIsOpen();
+
+    SWIGTYPE_p_p_char key = tiledb.new_charpp();
+    SWIGTYPE_p_unsigned_int key_len = tiledb.new_uintp();
+    SWIGTYPE_p_tiledb_datatype_t value_type = tiledb.new_tiledb_datatype_tp();
+    SWIGTYPE_p_unsigned_int value_num = tiledb.new_uintp();
+    SWIGTYPE_p_p_void value = tiledb.new_voidpArray(0);
+
+    ctx.handleError(
+        tiledb.tiledb_array_get_metadata_from_index(
+            ctx.getCtxp(), arrayp, index, key, key_len, value_type, value_num, value));
+
+    String keyString = tiledb.charpp_value(key);
+    long valueLength = tiledb.uintp_value(value_num);
+    Datatype nativeType = Datatype.fromSwigEnum(tiledb.tiledb_datatype_tp_value(value_type));
+
+    NativeArray result = new NativeArray(ctx, nativeType, value, (int) valueLength);
+
+    tiledb.delete_uintp(value_num);
+    tiledb.delete_uintp(key_len);
+    tiledb.delete_charpp(key);
+    tiledb.delete_tiledb_datatype_tp(value_type);
+
+    return new Pair<String, NativeArray>(keyString, result);
+  }
+
+  /**
+   * Checks if the key is present in the Array metadata. The array must be opened in READ mode,
+   * otherwise the function will error out.
+   *
+   * @param key a key to retrieve from the metadata key-value
+   * @return true if the key is present in the metadata, false if it is not
+   * @throws TileDBError A TileDB exception
+   */
+  public Boolean hasMetadataKey(String key) throws TileDBError {
+    checkIsOpen();
+
+    SWIGTYPE_p_tiledb_datatype_t value_type = tiledb.new_tiledb_datatype_tp();
+    SWIGTYPE_p_int has_key = tiledb.new_intp();
+
+    ctx.handleError(
+        tiledb.tiledb_array_has_metadata_key(ctx.getCtxp(), arrayp, key, value_type, has_key));
+
+    Boolean result = tiledb.intp_value(has_key) > 0;
+
+    tiledb.delete_intp(has_key);
+    tiledb.delete_tiledb_datatype_tp(value_type);
+
+    return result;
+  }
+
+  /**
+   * Puts a metadata key-value item to an open array. The array must be opened in WRITE mode,
+   * otherwise the function will error out.
+   *
+   * @param key a key to assign to the input value
+   * @param value the metadata to put into the Array metadata
+   * @throws TileDBError A TileDB exception
+   */
+  public void putMetadata(String key, NativeArray value) throws TileDBError {
```

Comment:
I'm wondering, should the following overload be a part of this PR

```java
public void putMetadata(String key, Object buffer, Class javaType) throws TileDBError {
  putMetadata(key, new NativeArray(ctx, buffer, javaType));
}
```

Or it can / should be discussed in a separate PR in case this one would pass.

Reply:
@pomadchin let's look into this as a followup PR. We wanted to initial changes merged, as they are a good foundation to build upon.

For the overloads, is this something you are interested in contributing? If not, @gsvic can add them, we just want to coordinate.

Reply:
Done! https://github.com/TileDB-Inc/TileDB-Java/pull/156